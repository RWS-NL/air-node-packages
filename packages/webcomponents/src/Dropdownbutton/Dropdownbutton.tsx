import ButtonGroup from '@material-ui/core/ButtonGroup';
import ClickAwayListener from '@material-ui/core/ClickAwayListener';
import Grid from '@material-ui/core/Grid';
import Grow from '@material-ui/core/Grow';
import MenuItem from '@material-ui/core/MenuItem';
import MenuList from '@material-ui/core/MenuList';
import Paper from '@material-ui/core/Paper';
import Popper from '@material-ui/core/Popper';
import ArrowDropDownIcon from '@material-ui/icons/ArrowDropDown';
import classnames from 'classnames';
import React, { FC, memo, MouseEvent as ReactMouseEvent, ReactNode, useRef, useState } from 'react';
import Button from '../Button/Button';
import css from './Dropdownbutton.scss';
import { cutText } from '@rws-air/utils';

export interface DropdownbuttonProps {
  /** The options to show in the dropdown menu */
  options: string[];
  /** Disable the entire dropdown button */
  disabled?: boolean;
  /** The icon displayed in the button to the right. This button is clicked to trigger {@link onClick()} */
  ButtonIcon: ReactNode;
  /** The maximum length of the text in the button before being cut off with ellipsis @default 30 */
  maxTextLength?: number;
  /** The default option to select, using zero-based index for {@link options} @default 0 */
  defaultOptionId?: number;
  /** Any options to disabled, using zero-based index for {@link options} */
  disabledOptionIds?: number[];
  /** The color type of the buttons @default primary */
  color?: 'primary' | 'secondary' | 'default' | 'inherit';
  /** The variant of the buttons @default contained */
  variant?: 'text' | 'outlined' | 'contained' | undefined;
  /** Data-qa tags to apply to the button group */
  buttonGroupDataQa?: string;
  /** Data-qa tags to apply to the dropdown button */
  dropdownButtonDataQa?: string;
  /** Data-qa tags to apply to the icon button */
  iconButtonDataQa?: string;
  /** Data-qa tags to apply to each menu item */
  menuItemDataQa?: string;
  /** Data-qa tags to apply to the popper */
  popperDataQa?: string;
  /** Custom CSS classes to pass to the dropdown button */
  dropdownButtonCustomClasses?: string | string[];
  /** Custom CSS classes to pass to the icon button */
  iconButtonCustomClasses?: string | string[];
  /** Custom CSS classes to pass to the buttonGroup */
  buttonGroupCustomClasses?: string | string[];
  /** The action to trigger when the button with {@link ButtonIcon} is pressed. Receives the currently selected option as first parameter */
  onClick(selectedOption: string): unknown;
}

/**
 * Constructs a button with dropdown menu options using pre-defined Rijkswaterstaat styling
 * @example
 * <Dropdownbutton ButtonIcon={<CloudDownload/>} onClick={() => undefined} options={[ 'one', 'two' ]} />
 */
export const Dropdownbutton: FC<DropdownbuttonProps> = props => {
  const [ open, setOpen ] = useState(false);
  const anchorRef = useRef<HTMLDivElement>(null);
  const [ selectedIndex, setSelectedIndex ] = useState(props.defaultOptionId || 0);

  const handleMenuItemClick = (_event: ReactMouseEvent<HTMLLIElement, MouseEvent>, index: number) => {
    setSelectedIndex(index);
    setOpen(false);
  };

  const handleToggle = () => {
    setOpen(prevOpen => !prevOpen);
  };

  const handleClose = (event: ReactMouseEvent<Document, MouseEvent>) => {
    if (anchorRef.current && anchorRef.current.contains(event.target as HTMLElement)) {
      return;
    }
    setOpen(false);
  };

  return (
    <Grid item container direction='column'>
      <Grid item xs={12}>
        <ButtonGroup
          variant={props.variant || 'contained'}
          color={props.color || 'primary'}
          ref={anchorRef}
          aria-label='split button'
          data-qa={props.buttonGroupDataQa}
          className={classnames(css.buttonGroup, props.buttonGroupCustomClasses)}
          classes={{ root: classnames({ [css.buttonGroupDisabled]: props.disabled }) }}
          disabled={props.disabled}
        >
          <Button
            variant={props.variant || 'contained'}
            color={props.color || 'primary'}
            size='small'
            aria-owns={open ? 'menu-list-grow' : undefined}
            aria-haspopup='true'
            onClick={handleToggle}
            startIcon={<ArrowDropDownIcon />}
            label={cutText(props.options[selectedIndex], props.maxTextLength || 30)}
            className={css.buttonOverwritesLeft}
            customlabelclasses={css.labelOverwrites}
            data-qa={props.dropdownButtonDataQa}
            customclasses={props.dropdownButtonCustomClasses}
          />
          <Button
            variant={props.variant || 'contained'}
            color={props.color || 'primary'}
            onClick={() => props.onClick(props.options[selectedIndex])}
            label={props.ButtonIcon}
            className={css.buttonOverwritesRight}
            data-qa={props.iconButtonDataQa}
            customclasses={props.iconButtonCustomClasses}
          />
        </ButtonGroup>
        <Popper
          open={open}
          data-qa={props.popperDataQa}
          anchorEl={anchorRef.current}
          transition
          placement='bottom-start'
          disablePortal
        >
          {({ TransitionProps, placement }) => (
            <Grow
              {...TransitionProps}
              style={{
                transformOrigin:
                  placement === 'bottom' ? 'center top' : 'center bottom',
              }}
            >
              <Paper id='menu-list-grow'>
                <ClickAwayListener onClickAway={handleClose}>
                  <MenuList>
                    {props.options.map((option, index) => (
                      <MenuItem
                        data-qa={`${props.menuItemDataQa}-${index}`}
                        key={option}
                        disabled={props.disabledOptionIds ? props.disabledOptionIds.includes(index) : false}
                        selected={index === selectedIndex}
                        onClick={event => handleMenuItemClick(event, index)}
                      >
                        {option}
                      </MenuItem>
                    ))}
                  </MenuList>
                </ClickAwayListener>
              </Paper>
            </Grow>
          )}
        </Popper>
      </Grid>
    </Grid>
  );
};

export default memo(Dropdownbutton);