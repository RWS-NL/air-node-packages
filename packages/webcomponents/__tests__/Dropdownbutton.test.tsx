import CloudDownload from '@material-ui/icons/CloudDownload';
import { Dropdownbutton, DropdownbuttonProps } from '@src/Dropdownbutton';
import { shallow, ShallowWrapper } from 'enzyme';
import React from 'react';

const mockButtonClick = jest.fn();

describe('Component Tests', () => {
  let dropdownButton: ShallowWrapper;

  beforeAll(() => {
    dropdownButton = shallow(
      <Dropdownbutton
        ButtonIcon={<CloudDownload />}
        onClick={mockButtonClick}
        options={['one', 'two']}
        menuItemDataQa='menu-item'
        iconButtonDataQa='icon-button'
        dropdownButtonDataQa='dropdown-button'
      />
    );
  });

  test('button function called onClick', () => {
    const iconButton = dropdownButton.find('[data-qa="icon-button"]');

    iconButton.simulate('click');

    expect(mockButtonClick).toHaveBeenCalledWith('one');
    expect(mockButtonClick).toHaveBeenCalledTimes(1);
  });
});

describe('Component Test - Disabled Button', () => {
  let dropdownButton: ShallowWrapper<DropdownbuttonProps>;

  beforeAll(() => {
    dropdownButton = shallow<DropdownbuttonProps>(
      <Dropdownbutton
        disabled
        ButtonIcon={<CloudDownload />}
        onClick={mockButtonClick}
        buttonGroupDataQa='button-group'
        options={['one', 'two']}
      />
    );
  });

  test('button should be disabled and no call should be made to the click function', () => {
    const buttonGroup = dropdownButton.find('[data-qa="button-group"]');
    expect(buttonGroup.prop('disabled')).toBe(true);
  });
});

describe('Snapshot Testing', () => {
  test('Required Props', () => {
    const dropdownButton = shallow(
      <Dropdownbutton ButtonIcon={<CloudDownload />} onClick={mockButtonClick} options={['one', 'two', 'three']} />
    );
    expect(dropdownButton).toMatchSnapshot();
  });

  test('Optional Props', () => {
    const dropdownButton = shallow(
      <Dropdownbutton
        ButtonIcon={<CloudDownload />}
        onClick={mockButtonClick}
        options={['one', 'two', 'three']}
        disabledOptionIds={[2]}
        color='secondary'
        variant='outlined'
        disabled
      />
    );
    expect(dropdownButton).toMatchSnapshot();
  });
});
