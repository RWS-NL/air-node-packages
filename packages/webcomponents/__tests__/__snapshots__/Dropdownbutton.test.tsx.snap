// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Testing Optional Props 1`] = `
<WithStyles(ForwardRef(Grid))
  container={true}
  direction="column"
  item={true}
>
  <WithStyles(ForwardRef(Grid))
    item={true}
    xs={12}
  >
    <WithStyles(ForwardRef(ButtonGroup))
      aria-label="split button"
      className="buttonGroup"
      classes={
        Object {
          "root": "buttonGroupDisabled",
        }
      }
      color="secondary"
      disabled={true}
      variant="outlined"
    >
      <Memo()
        aria-haspopup="true"
        className="buttonOverwritesLeft"
        color="secondary"
        customlabelclasses="labelOverwrites"
        label="one"
        onClick={[Function]}
        size="small"
        startIcon={<UNDEFINED />}
        variant="outlined"
      />
      <Memo()
        className="buttonOverwritesRight"
        color="secondary"
        label={<UNDEFINED />}
        onClick={[Function]}
        variant="outlined"
      />
    </WithStyles(ForwardRef(ButtonGroup))>
    <ForwardRef(Popper)
      anchorEl={null}
      disablePortal={true}
      open={false}
      placement="bottom-start"
      transition={true}
    >
      <Component />
    </ForwardRef(Popper)>
  </WithStyles(ForwardRef(Grid))>
</WithStyles(ForwardRef(Grid))>
`;

exports[`Snapshot Testing Required Props 1`] = `
<WithStyles(ForwardRef(Grid))
  container={true}
  direction="column"
  item={true}
>
  <WithStyles(ForwardRef(Grid))
    item={true}
    xs={12}
  >
    <WithStyles(ForwardRef(ButtonGroup))
      aria-label="split button"
      className="buttonGroup"
      classes={
        Object {
          "root": "",
        }
      }
      color="primary"
      variant="contained"
    >
      <Memo()
        aria-haspopup="true"
        className="buttonOverwritesLeft"
        color="primary"
        customlabelclasses="labelOverwrites"
        label="one"
        onClick={[Function]}
        size="small"
        startIcon={<UNDEFINED />}
        variant="contained"
      />
      <Memo()
        className="buttonOverwritesRight"
        color="primary"
        label={<UNDEFINED />}
        onClick={[Function]}
        variant="contained"
      />
    </WithStyles(ForwardRef(ButtonGroup))>
    <ForwardRef(Popper)
      anchorEl={null}
      disablePortal={true}
      open={false}
      placement="bottom-start"
      transition={true}
    >
      <Component />
    </ForwardRef(Popper)>
  </WithStyles(ForwardRef(Grid))>
</WithStyles(ForwardRef(Grid))>
`;
